src/test/run-fail/cstring-drop.rs

```diff
@@ -0,0 +1,35 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Test that `CString::new("hello").unwrap().as_ptr()` pattern
+// leads to failure. Technically, this test relies on undefined
+// behavior, but let's hope we get an assertion error and not a
+// random segfault.
+
+//error-pattern:assertion failed
+
+use std::ffi::{CString, CStr};
+use std::os::raw::c_char;
+
+fn main() {
+    // Repeat ten times and check that at least once we get "hello" back.
+    // If we fail every time, this means that `impl Drop for CString`
+    // correctly protects us.
+    let xs = vec![CString::new("Hello").unwrap(); 10];
+    let ys = xs.iter().map(|s| s.as_ptr()).collect::<Vec<_>>();
+    drop(xs);
+    assert!(ys.into_iter().any(is_hello));
+}
+
+fn is_hello(s: *const c_char) -> bool {
+    let s = unsafe { CStr::from_ptr(s) };
+    let hello = CString::new("Hello").unwrap();
+    s == hello.as_ref()
```

Comment:
I think this is technically undefined behavior, so perhaps this test could be isolated into a separate process? That is, it's fine if the process segfaults or otherwise aborts abnormally, but it's *not* fine for it to return successfully.